crypto/x509/x509_vfy.c

```diff
@@ -3394,21 +3352,16 @@ static int build_chain(X509_STORE_CTX *ctx)
     case X509_TRUST_UNTRUSTED:
     default:
         num = sk_X509_num(ctx->chain);
-        if (num > depth)
-            return verify_cb_cert(ctx, NULL, num-1,
-                                  X509_V_ERR_CERT_CHAIN_TOO_LONG);
-        if (DANETLS_ENABLED(dane) &&
-            (!DANETLS_HAS_PKIX(dane) || dane->pdpth >= 0))
-            return verify_cb_cert(ctx, NULL, num-1, X509_V_ERR_DANE_NO_MATCH);
-        if (self_signed && sk_X509_num(ctx->chain) == 1)
-            return verify_cb_cert(ctx, NULL, num-1,
-                                  X509_V_ERR_DEPTH_ZERO_SELF_SIGNED_CERT);
-        if (self_signed)
-            return verify_cb_cert(ctx, NULL, num-1,
-                                  X509_V_ERR_SELF_SIGNED_CERT_IN_CHAIN);
-        if (ctx->num_untrusted < num)
-            return verify_cb_cert(ctx, NULL, num-1,
-                                  X509_V_ERR_UNABLE_TO_GET_ISSUER_CERT);
+        CHECK_CB(num > depth, ctx, NULL, num-1, X509_V_ERR_CERT_CHAIN_TOO_LONG);
+        CHECK_CB(DANETLS_ENABLED(dane)
+                     && (!DANETLS_HAS_PKIX(dane) || dane->pdpth >= 0),
+                 ctx, NULL, num-1, X509_V_ERR_DANE_NO_MATCH);
+        CHECK_CB(self_signed && sk_X509_num(ctx->chain) == 1,
+                 ctx, NULL, num-1, X509_V_ERR_DEPTH_ZERO_SELF_SIGNED_CERT);
+        CHECK_CB(self_signed,
```

Comment:
This change means that if the verify cb decides to ignore the X509_V_ERR_DEPTH_ZERO_SELF_SIGNED_CERT error, it will still be again called with X509_V_ERR_SELF_SIGNED_CERT_IN_CHAIN - which I believe is not correct. And I do not know if the following condition makes sense in this situation either.

Reply:
I had simply assumed that the error conditions in that code section are independent of each other, such that it would make sense to report any and all of them.
Yet good point that for the conditions you refer to here this is not really the case.
So I've improved the calls to the new `CHECK_CB()` macro accordingly.